### DISCUSSION TITLE
Console and non-console version, smarter way?

### DISCUSSION THREAD

**Discussion by Safihre** [pyinstaller/pyinstaller#8006] [Help] [answered]
On Windows, we need a version of our exe that has a console and one that doesn't. 
From the Issues here I know there is no runtime-way to do that, so we just have to create 2 applications.
We currently do it like this:
https://github.com/sabnzbd/sabnzbd/blob/develop/builder/SABnzbd.spec#L110-L150

However, this creates also 2 _identical_ `.pkg` files of 5MB each. 
Luckily the compression algorithm of NSIS installer detects the data is identical, so it can "compress" them together.
Still, I wonder if there is a more space-efficient way?
I noticed I could also not bundle things in the `.pkg`, is there a downside to that?

**Comment by rokm** [comment, marked as answer]
There is no smarter way, unfortunately. In onedir builds, PKG contains the bootstrap modules, run-time hooks, the bootloader options (if any), the entry-point script, and the PYZ archive. And the latter is probably the most prominent contributor to the PKG size.

Eventually, I plan to replace the current "multi-package" mess of the MERGE with an option for executables to share a single external PYZ (which would help multiple onedir executables bundled together) and an option to share a single external PKG for onefile (with data, binaries, and PYZ; the bootstrap modules, run-time hooks, bootloader options and entry-point scripts remain in executable-embedded PYZ). Then we can retire the MERGE, and possibly also the .pkg side-loading mode.


> However, this creates also 2 _identical_ `.pkg` files of 5MB each.
> Luckily the compression algorithm of NSIS installer detects the data is identical, so it can "compress" them together.
> Still, I wonder if there is a more space-efficient way?

Not on Windows. On other OSes, you could probably remove a copy and create a symlink.

> I noticed I could also not bundle things in the `.pkg`, is there a downside to that?

Actually, .pkg sideloading (`append_pkg=False`) is the non-default. But it makes no difference in this case, because the PKG would be embedded in the executable, so the duplication would just be shifted to the two executables.

>On Windows, we need a version of our exe that has a console and one that doesn't. From the Issues here I know there is no runtime-way to do that, so we just have to create 2 applications.

Depending on what exactly you want to achieve, you could now try faking dynamic console with [`--hide-console`](https://pyinstaller.org/en/stable/usage.html#cmdoption-hide-console`) (and equivalent `hide_console` argument to the `EXE`).

**Reply by Safihre** [reply]
Thanks so much! 
`--hide-console` was exactly what we needed!
On macOS we already instruct our users to just add `--console` if they want to run from command line with console output.
